spanner/value.go

```diff
@@ -2069,6 +2072,47 @@ func decodeValue(v *proto3.Value, t *sppb.Type, ptr interface{}, opts ...decodeO
 			return decodableType.decodeValueToCustomType(v, t, acode, atypeAnnotation, ptr)
 		}
 
+		// Check if the proto encoding is for an array of proto message or array of proto enum.
+		// The below scenario cannot be handled in a switch case statement because the pointer in this case
+		// would be an array of proto message or array of proto enum which are user defined types.
+		// Adding a generic switch case for these user defined types is not possible, hence we use proto encoding.
+		if code == sppb.TypeCode_ARRAY && (acode == sppb.TypeCode_PROTO ||
+			acode == sppb.TypeCode_ENUM || acode == sppb.TypeCode_INT64 ||
+			acode == sppb.TypeCode_BYTES) {
```

Comment:
I don't think we should go into this branch for `INT64` and `BYTES`. As far as I can see, it does work in this case (that is; it does not interfere with any other cases), but it feels dangerous. We are basically saying that the 'catch-all' case for `ARRAY<INT64>` is `ProtoEnum` and for `ARRAY<BYTES>` is `Proto`. That is a choice that we cannot undo in the future.

Reply:
@olavloite Here we do not 'catch-all' case for `ARRAY<INT64>` as `ProtoEnum` and for `ARRAY<BYTES>` as `ProtoMessage`. It depends on the interface that is passed by the user for decoding.

Let me try to explain this scenario for `ARRAY<INT64>`. Let us assume that spanner backend returns `code` as  `sppb.TypeCode_ARRAY`and `acode` as `sppb.TypeCode_INT64`.

**Scenario 1:**
If the user passes `[]int64` as the interface to read data then it gets caught in `decodeValue` function way before in the `switch` case as pointed [here](https://github.com/googleapis/google-cloud-go/blob/33419f9f5267aa67e8b240134251c13d074cfd58/spanner/value.go#L1363).
**Scenario 2:**
Similarly, if the user passes an `[]ProtoEnum` as the interface to read data then it executes the above [code](https://github.com/googleapis/google-cloud-go/blob/33419f9f5267aa67e8b240134251c13d074cfd58/spanner/value.go#L2079) as the interface passed does not satisfy any case in `switch` condition and comes down here.

I have added integration tests for these scenarios in [Scenario 1](https://github.com/googleapis/google-cloud-go/blob/33419f9f5267aa67e8b240134251c13d074cfd58/spanner/integration_test.go#L1854) and [Scenario 2](https://github.com/googleapis/google-cloud-go/blob/33419f9f5267aa67e8b240134251c13d074cfd58/spanner/integration_test.go#L2123)

Reply:
Yes, I understand that, so it's not a full catch-all case for `ARRAY<INT64>`, but it is for everything __except__ `[]int64`. That means that if we in the future have some other type that we want to map to `ARRAY<INT64>` with a specific Go type, then we have a problem. Because that would then introduce a breaking change, as it would then no longer go into this branch, but into the branch for the specific type.

Reply:
@olavloite 
I have modified the decoding logic for proto array by checking the interface passed instead of checking the proto encoding passed from spanner backend - [Link](https://github.com/googleapis/google-cloud-go/blob/0ea08c1b557aff2aa32add463422c340f0c2349e/spanner/value.go#L2078)
This will not cause a breaking change as we go into the branch only if interface passed is of type array of proto columns. Can you please verify this?

Reply:
Thanks, LGTM